code/game/machinery/crusher_piston.dm

```diff
@@ -0,0 +1,328 @@
+	asmtype = "standalone"
+	icon_state = "standalone"
+	return
+
+/obj/machinery/crusher_base/process()
+	if (process_lock)
+		log_debug("crusher_piston process() has been called while it was still locked. Aborting")
+		return
+	process_lock = 1
+	var/timediff = (world.time - action_start_time) / 10
+
+	//Check what action should be performed
+	if(action == "idle")
+		action_start_time = world.time
+		initial = 1
+	else if(action == "extend")
+		//If we are idle, flash the warning lights and then put us into pre_start once we are done
+		if(status == "idle")
+			if(initial)
+				playsound(loc, 'sound/machines/airalarm.ogg', 50, 1)	//Plays a beep
+				initial = 0
+			//TODO: Flash the lights, Sound the alarm
+			if(timediff > time_stage_pre)
+				initial = 1
+				status = "pre_start" // Bring us into pre start
+				action_start_time = world.time
+				
+
+		//We are now ready to expand the first piston
+		else if(status == "pre_start")
+			//Call extend on all the stage 1 pistons
+			if(initial)
+				initial = 0
+				for(var/obj/machinery/crusher_piston/pstn in pistons)
+					pstn.extend_0_1()
+			if(timediff > time_stage_1)
+				status = "stage1"
+				action_start_time = world.time
+				initial = 1
+		
+		//Extend the second piston
+		else if(status == "stage1")
+			//Call extend on all the stage 2 pistons
+			if(initial)
+				initial = 0
+				for(var/obj/machinery/crusher_piston/pstn in pistons)
+					pstn.extend_1_2()
+			if(timediff > time_stage_2)
+				status = "stage2"
+				action_start_time = world.time
+				initial = 1
+
+		//Extend the thrid piston
+		else if(status == "stage2")
+			//Call extend on all the stage 2 pistons
+			if(initial)
+				initial = 0
+				for(var/obj/machinery/crusher_piston/pstn in pistons)
+					pstn.extend_2_3()
+			if(timediff > time_stage_3)
+				status = "stage3"
+				action_start_time = world.time
+				initial = 1
+		
+		//Wait a moment, then reverse the direction
+		else if(status == "stage3")
+			if(initial)
+				initial = 0
+			if(timediff > time_stage_full)
+				action = "retract"
+				action_start_time = world.time
+				initial = 1
+	
+	//Retract the pistons
+	else if(action == "retract")
+		//Retract the 3rd stage pistons
+		if(status == "stage3")
+			if(initial)
+				initial = 0
+				for(var/obj/machinery/crusher_piston/pstn in pistons)
+					pstn.retract_3_0()
+			if(timediff > time_stage_1) //Once the time is up, reset the icon state
+				for(var/obj/machinery/crusher_piston/pstn in pistons)
+					pstn.icon_state = initial(pstn.icon_state)
+				status = "idle"
+				action = "idle"
+				action_start_time = world.time
+				initial = 1
+	
+	//Move all the items in the move list
+	for(var/atom/movable/AM in items_to_move)
+		if(!AM.simulated)
+			items_to_move -= AM
+			continue
+		if(istype(AM,/obj/machinery/crusher_piston))
+			items_to_move -= AM
+			continue
+		items_to_move -= AM
+		//If a item could not be moved. Add it to the crush list
+		if(!AM.piston_move())
+			items_to_crush += AM
+		CHECK_TICK
```

Comment:
Please add `set waitfor = FALSE` to this proc if you're CHECK_TICKing it, CHECK_TICK sleeps and process() must not sleep without being spawn()'d or waitfor'd.